packages/sil-governance/src/routes/dashboard/components/ProposalsList.tsx

```diff
@@ -34,14 +48,47 @@ const getFilter = (filterType: ElectionFilter, currentUser: Address | null): Pro
   }
 };
 
-function compareById<E extends { readonly id: number }>(element1: E, element2: E): number {
-  return element1.id - element2.id;
-}
+type ProposalsComparator = (proposal1: ProposalProps, proposal2: ProposalProps) => number;
+
+const compareByIdDescending: ProposalsComparator = (proposal1, proposal2): number => {
+  return proposal1 > proposal2 ? 1 : -1;
+};
+const compareByExpiryDate: ProposalsComparator = (proposal1, proposal2): number => {
+  return proposal1.expiryDate.getTime() - proposal2.expiryDate.getTime();
+};
+const compareByStartDate: ProposalsComparator = (proposal1, proposal2): number => {
+  return proposal1.startDate.getTime() - proposal2.startDate.getTime();
+};
+const compareByVote: ProposalsComparator = (proposal1, proposal2): number => {
+  if (proposal2.vote === undefined) return -1;
+  if (proposal1.vote === undefined) return 1;
+  return proposal1.vote - proposal2.vote;
+};
 
 const ProposalsList = ({ filterType }: Props): JSX.Element => {
-  const proposals = ReactRedux.useSelector((state: RootState) => state.proposals.proposals);
-  const governor = ReactRedux.useSelector((state: RootState) => state.extension.governor);
-  const blockchain = ReactRedux.useSelector((state: RootState) => state.blockchain);
+  const proposals = useSelector((state: RootState) => state.proposals.proposals);
+  const governor = useSelector((state: RootState) => state.extension.governor);
+  const blockchain = useSelector((state: RootState) => state.blockchain);
+
+  const [comparator, setComparator] = useState<ProposalsComparator>(() => compareByIdDescending);
+
+  const changeComparator = (selectedItem: SelectFieldItem): void => {
+    switch (selectedItem.name) {
+      case ComparatorLabels.ExpiryDate:
+        setComparator(() => compareByExpiryDate);
+        break;
+      case ComparatorLabels.StartDate:
+        setComparator(() => compareByStartDate);
+        break;
+      case ComparatorLabels.Vote:
+        setComparator(() => compareByVote);
+        break;
+      default:
+        setComparator(() => compareByIdDescending);
```

Comment:
Can't you use simply `setComparator(compareByIdDescending);`? Or will that be interpreted as a state update callback function?

Reply:
This gave me a lot of headache. If you pass the comparator directly it gets executed immediately and therefore the app crashes because the `proposal1` and `proposal2` arguments are `undefined`.

The same happens in `useState<ProposalsComparator>(() => compareByIdDescending);`

Reply:
I don't think it is because the proposals are undefined. Even if the list of proposals was not ready yet, you get an empty list that can be sorted.

I think what happens is the following: Let look at the initial call `const [comparator, setComparator] = useState<ProposalsComparator>(() => compareByIdDescending);`. `useState` has the following API:

```ts
    /**
     * Returns a stateful value, and a function to update it.
     *
     * @version 16.8.0
     * @see https://reactjs.org/docs/hooks-reference.html#usestate
     */
    function useState<S>(initialState: S | (() => S)): [S, Dispatch<SetStateAction<S>>];
```

So you can either pass in a value or an update function that returns this value (see https://reactjs.org/docs/hooks-reference.html#usestate). At runtime, if the value you pass in is any kind of function, it gets executed and the retuned  value is used: https://github.com/facebook/react/blob/v16.10.0/packages/react-reconciler/src/ReactFiberHooks.js#L816-L818

So generally speaking it is probably not a good idea to store the comparators directly in the state. It would be better to use some kind of sort identifier and then switch over the identifier in

```ts
    .filter(filter)
    .sort(comparator);
```